src/coalesce.c

```diff
@@ -0,0 +1,155 @@
+#include "data.table.h"
+
+SEXP coalesce(SEXP x, SEXP values) {
+  int JJ = LENGTH(values), nprotect = 0, NN = LENGTH(x);
+  
+  // pass over x in parallel to count NA
+  int na_count = 0;
+  switch(TYPEOF(x)) {
+  case LGLSXP: {
+    int *restrict x_ptr = LOGICAL(x);
+    #pragma omp parallel for schedule(dynamic) reduction(+:na_count)
+    for (int i = 0; i < NN; i++) na_count += (int) (x_ptr[i] == NA_LOGICAL);
+  } break;
+  case INTSXP: {
+    int *restrict x_ptr = INTEGER(x);
+    #pragma omp parallel for schedule(dynamic) reduction(+:na_count)
+    for (int i = 0; i < NN; i++) na_count += (int) (x_ptr[i] == NA_INTEGER);
+  } break;
+  case REALSXP: {
+    double *restrict x_ptr = REAL(x);
+    #pragma omp parallel for schedule(dynamic) reduction(+:na_count)
+    for (int i = 0; i < NN; i++) na_count += (int) (ISNA(x_ptr[i]));
+  } break;
+  case STRSXP: {
+    for (int i = 0; i < NN; i++) na_count += (int) (STRING_ELT(x, i) == NA_STRING);
+  } break;
+  default: error("Incompatible type");
+  }
+  
+  switch(TYPEOF(x)) {
+  case LGLSXP: {
+    int *x_ptr = LOGICAL(x);
+    
+    SEXP out = PROTECT(allocVector(LGLSXP, NN)); nprotect++;
+    int *out_ptr = LOGICAL(out);
+    
+    for (int i = 0; i < NN; i++) {
+      if (na_count > 0 && x_ptr[i] == NA_LOGICAL) {
+        int j = 0;
+        while (j < JJ) {
+          SEXP this_j = VECTOR_ELT(values, j);
+          int take_idx;
+          if (LENGTH(this_j) == 1) {
+            take_idx = 0;
+          } else {
+            take_idx = i;
+          }
+          int this_replacement = LOGICAL(this_j)[take_idx];
```

Comment:
`LOGICAL(this_j)[take_idx]` -> `LOGICAL(this_j)[LENGTH(this_j) == 1 ? 0 : i]` and we can remove 5 lines above